### PR TITLE
 Added a .NET Core version of the Azure DevOps (vsts) build/release task.

### PR DESCRIPTION
**Pull request by yohanb**
Hi,
Here's a PR for the following request change:
https://github.com/GitTools/GitVersion/issues/1496

I create a separate task specifically for .NET Core so it doesn't interfere with the current portable version.
I tested the task on my personal Azure DevOps account on a hosted Ubuntu agent and it works!  It requires for .NET Core 2.0+ to be installed on the agent.

Let me know if I can do anything else to get this thing out in the wild.

Cheers.

### CODE REVIEW

**Comment by arturcic** [comment]
Thank you for the contribution. That will be a great step to support the Unix agents on Azure DevOps.

**Comment by yohanb** [comment]
@arturcic the build steps have failed because I added a new `VsixNetCoreOutputFilePath` parameter.

File: build.cake [hunk 1]
```diff
@@ -238,6 +238,11 @@ Task("Copy-Files")
     CopyFileToDirectory(portableDir + "/" + "GitVersion.exe", tfsPath);
     CopyDirectory(portableDir.Combine("lib"), tfsPath.Combine("lib"));
 
+    // Vsix dotnet core
+    var tfsNetCorePath = new DirectoryPath("./src/GitVersionTfsTask.NetCore/GitVersionTask");
```
[the remark this answers is not in the document]
Comment (yohanb) [reply build.cake]: OK, I'll find the time to rework the folders this week.

[the remark this answers is not in the document]
Comment (arturcic) [reply build.cake]: Thanks, you can have as inspiration the way [here](https://github.com/Microsoft/vsts-extension-build-release-tasks/tree/master/BuildTasks) it is structured

**Comment by yohanb** [comment]
hey @arturcic ,
I modified the code as you requested.  I haven't really validate it works besides the cake build and packaging.
Keep me posted!

**Comment by arturcic** [comment]
@yohanb, I'll have a look soon, thanks

**Comment by mitchdenny** [comment]
Would love to see this merged :)

**Comment by arturcic** [comment]
@yohanb  Can you please make the changes so that we can move forward?  We should also consider using azure-pipelines-task-lib npm package as vsts-task-lib is deprecated [details](https://www.npmjs.com/package/vsts-task-lib)

File: src/GitVersionTfsTask/GitVersionTask/task.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
   "version": {
     "Major": "4",
     "Minor": "0",
-    "Patch": "0"
+    "Patch": "1"
```
Comment (arturcic) [inline src/GitVersionTfsTask/GitVersionTask/task.json]: Better to avoid committing these changes as they are changed on every build

**Comment by yohanb** [comment]
@arturcic done making the changes.  Only thing I havent validated is the creation of separate vsix packages.

**Comment by arturcic** [comment]
Looks great! I have run the build locally and it seems the paths to the vsix (the ones used for publishing) differ from the actual paths

image

Can you update the path here? https://github.com/GitTools/GitVersion/pull/1499/files#diff-faf8aebe07f0c09fe65d5a4be490bb83R43

File: build/paths.cake [hunk 3]
```diff
@@ -40,6 +40,7 @@ public class BuildPaths
         var releaseNotesOutputFilePath = buildArtifactDir.CombineWithFilePath("releasenotes.md");
 
         var vsixOutputFilePath = buildArtifactDir.CombineWithFilePath("gittools.gitversion-" + semVersion + ".vsix");
+        var vsixNetCoreOutputFilePath = buildArtifactDir.CombineWithFilePath("gittools.gitversion_netcore-" + semVersion + ".vsix");
```
Comment (arturcic) [inline build/paths.cake]: The path here should match the artifact's path for publishing

Comment (arturcic) [reply build/paths.cake]: basically just change `gitversion_netcore` with `gitversion-netcore`

Comment (yohanb) [reply build/paths.cake]: nice find!  Done and pushed.

**Comment by yohanb** [comment]
> there is one missing line you should add so that the new vsix file is published to AppVeyor and to the marketplace https://github.com/GitTools/GitVersion/blob/master/build/parameters.cake#L106, Please add the netcore path as well.

done

**Comment by arturcic** [comment]
@yohanb Looks great to me. Can we prepare it for merge by squashing all the commits as one commit? Then you can rebase it on master, to keep the commit history clean? Thanks

**Comment by yohanb** [comment]
@arturcic do you mean rebase on my forked master and create another PR for yohanb:master -> gittools:master ?

**Comment by arturcic** [comment]
> @arturcic do you mean rebase on my forked master and create another PR for yohanb:master -> gittools:master ?

looks like a simple squash is not possible as you have merged the master already, So all the commits after the master merge commit can be squashed. Need to think about

**Comment by yohanb** [comment]
> > @arturcic do you mean rebase on my forked master and create another PR for yohanb:master -> gittools:master ?
> 
> looks like a simple squash is not possible as you have merged the master already, So all the commits after the master merge commit can be squashed. Need to think about

it should be possible to simply squash and commit the PR according to the docs:
https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits

**Comment by arturcic** [comment]
> it should be possible to simply squash and commit the PR according to the docs:
> https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits

ok then, give it a try.

**Comment by yohanb** [comment]
> > it should be possible to simply squash and commit the PR according to the docs:
> > https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits
> 
> ok then, give it a try.

can't
image

**Comment by yohanb** [comment]
@arturcic everything is clean and rebased on my forked master.  How do you wish to commit the changes upstream to this repo?

**Comment by arturcic** [comment]
@yohanb first of all you should push (**forced**) to your origin, this way it will appear on your origin master branch and it will automatically appear here. So the history will be overwritten.

**Comment by yohanb** [comment]
> @yohanb first of all you should push (**forced**) to your origin, this way it will appear on your origin master branch and it will automatically appear here. So the history will be overwritten.

ok thanks, I forgot to force push on this feature branch.  Not used to rewriting git history 🤣 
here goes nothing!

**Comment by arturcic** [comment]
@yohanb. I have fixed the rebase conflict and ran the build. The changes are ready for merging. I'll merge the PR. Thanks for the contribution and the patience.

**Comment by jmezach** [comment]
Awesome work guys!

**Comment by yohanb** [comment]
> @yohanb. I have fixed the rebase conflict and ran the build. The changes are ready for merging. I'll merge the PR. Thanks for the contribution and the patience.

My pleasure.  Hopefully this will enable greater adoption of GitVersion no matter the platform!

**Comment by jmezach** [comment]
@arturcic Looks like the new extension version isn't yet available on the Marketplace. Is there a manual step still involved to make that happen?

**Comment by arturcic** [comment]
that's right, and that's because the marketplace does not allow to publish pre-release version, and we'll get a new version on marketplace when we have a stable version released

**Comment by jmezach** [comment]
I see. Do you have an idea when that will happen? We could greatly benefit from this since we would no longer have to delay our macOS and Linux builds for the Windows builds to finish.

**Comment by arturcic** [comment]
As far as I know we don't have a well defined milestones structure with issues that need to be included in one release or another. I guess we need that, @gep13 , @JakeGinnivan , @asbjornu - do you have anything to add?

**Comment by jmezach** [comment]
Perhaps a preview version of the Azure DevOps extension could be useful as well. I'm developing an open source extension myself as well and from our CI pipeline we build and package the preview version of the extension and we just monotonically increase the patch version number. Then when we ship we use the GitVersion generated version as the version number. This allows us to dogfood the extension ourselves and we could even share the preview extension with specific accounts if we want.

**Comment by asbjornu** [comment]
We don't have a well defined process for stable releases, no. Is it so that this particular PR is impossible to actually test without it being published to Azure DevOps, which we can't do until it has a stable version number?

**Comment by arturcic** [comment]
It is possible to test locally the tasks as they are nodejs apps, but to check how they work on Azure DevOps as part of the build you need to install it as far as I know

**Comment by jmezach** [comment]
@arturcic @asbjornu Yes, you can only test an Azure DevOps extension by uploading it to the Marketplace and as far as I know it only supports major.minor.patch versions and no -prerelease tags. 

However as I mentioned you can have a separate extension (with a different id in the vss-extension.json) that can be marked as preview and can even be private (as in, only available in your own account or shared with specific accounts). You can then simply increment the patch version every time you build and upload the extension automatically. Then when you're ready to ship you can change the ID of the extension to the original one and use the appropriate version number.

I'm currently using [this extension](https://marketplace.visualstudio.com/items?itemName=ms-devlabs.vsts-developer-tools-build-tasks) to automate that inside of my CD pipeline.

**Comment by arturcic** [comment]
@jmezach  thanks for suggestion. I'll think about this, and probably we'll have a separate extension for pre-release - something like GitVersion-beta or similar. 

> However as I mentioned you can have a separate extension (with a different id in the vss-extension.json) that can be marked as preview and can even be private 

Are you talking about [this](https://docs.microsoft.com/en-us/azure/devops/extend/develop/manifest?view=vsts#mark-an-extension-public) ?

**Comment by jmezach** [comment]
So the way that I've set this up currently is the the [vss-extension.json](https://github.com/Augurk/vsts-extension/blob/feature/xplat-commandline/src/vss-extension.json) in our source repository uses the ID augurk.augurk-preview. We process that through our [CI pipeline](https://github.com/Augurk/vsts-extension/blob/feature/xplat-commandline/azure-pipelines.yml) to generate a .vsix with the -preview tag. We upload that .vsix as an artifact associated with a particular build.

Then in our [release pipeline](https://dev.azure.com/augurk/Augurk/_releaseProgress?_a=release-pipeline-progress&releaseId=100&view=mine) (not sure if you can actually see this, but I hope so ;)) we've defined two stages: staging and production. In the staging stage we use tasks from the extension I mentioned earlier to determine the version of the preview extension that's currently on the Marketplace. We then increase the patch number, repackage the VSIX and upload it to the Marketplace. Once that's done we are able to try out the extension within our own Azure DevOps account, but it's not visible to other accounts unless we explicitly share it with them.

When we're ready to ship (which usually means merging the changes to the master branch) we process it through the same pipeline and then promote it to production. In that stage we repackage the extension and remove the -preview tag and also stamp the correct version number (as generated by GitVersion) into it. After that anyone can use the new version of the extension.

@arturcic Yes, that is exactly what I'm talking about. In fact, you could even have a stage where the extension is just private and marked as preview, then a public version that's marked as preview and a public version that is the released version. That's not what we're doing, but it's definitely an option.

**Comment by arturcic** [comment]
Thanks for explanation. I guess we can have a `GitVersion-preview` extension that will be public and in preview, and it will be updated on every pre-release build. On stable release we'll  pack the extension as `GitVersion` as it is now. Same for the second extension - the one running on .net core.

**Comment by jmezach** [comment]
Sounds like a great plan. Looking forward to trying it out.

**Comment by asbjornu** [comment]
@arturcic: That sounds like a great plan. How do we implement it? Do you have the required time and expertise to do it?

**Comment by arturcic** [comment]
I don't have much time, but will try to find some. The implementation is just in the build script - we check if it's stable release or pre-release, and we add the suffix or not in the vss-extension.json. Basically we change this json file depending on the type of the release.

**Comment by yohanb** [comment]
happy to see things moving here! 👍

**Pull request merged**